crypto/rand/rand_unix.c

```diff
@@ -119,26 +133,92 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 #   error "Seeding uses urandom but DEVRANDOM is not configured"
 #  endif
 
+#  if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
+#   if __GLIBC_PREREQ(2, 25)
+#    define OPENSSL_HAVE_GETRANDOM
+#   endif
+#  endif
+
+#  if (defined(__FreeBSD__) && __FreeBSD__ >= 12)
+#   define OPENSSL_HAVE_GETRANDOM
+#  endif
+
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+#   include <sys/random.h>
+#  endif
+
 #  if defined(OPENSSL_RAND_SEED_OS)
 #   if !defined(DEVRANDOM)
 #    error "OS seeding requires DEVRANDOM to be configured"
 #   endif
 #   define OPENSSL_RAND_SEED_DEVRANDOM
-#   if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
-#    if __GLIBC_PREREQ(2, 25)
-#     define OPENSSL_RAND_SEED_GETRANDOM
-#    endif
-#   endif
-#  endif
-
-#  ifdef OPENSSL_RAND_SEED_GETRANDOM
-#   include <sys/random.h>
 #  endif
 
 #  if defined(OPENSSL_RAND_SEED_LIBRANDOM)
 #   error "librandom not (yet) supported"
 #  endif
 
+#  if (defined(__FreeBSD__) && __FreeBSD__ >= 11)
+/*
+ * sysctl_random(): Use sysctl() to read a random number from the kernel
+ * Returns the size on success, 0 on failure.
+ */
+static size_t sysctl_random(void *buf, size_t buflen)
+{
+    int mib[2];
+    size_t done = 0;
+    size_t len;
+
+    mib[0] = CTL_KERN;
+    mib[1] = KERN_ARND;
+
+    do {
+        len = buflen;
+        if (sysctl(mib, 2, buf, &len, NULL, 0) == -1)
+            return done;
+        done += len;
+        buf += len;
+        buflen -= len;
+    } while (buflen > 0);
+
+    return done;
+}
+#  endif
+
+/*
+ * syscall_random(): Try to get random data using a system call
+ * returns the number of bytes returned in buf, or <= 0 on error.
+ */
+int syscall_random(void *buf, size_t buflen)
+{
+#  if defined(OPENSSL_HAVE_GETRANDOM)
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
+#   endif
+    return (int)getrandom(buf, buflen, 0);
+#  endif
+#  if defined(__linux) && defined(SYS_getrandom)
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
+#   endif
+    return (int)syscall(SYS_getrandom, buf, buflen, 0);
+#  endif
+#  if (defined(__FreeBSD__) && __FreeBSD__ >= 11)
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
+#   endif
+    return (int)sysctl_random(buf, buflen);
+#  endif
+   /* Supported since OpenBSD 5.6 */
+#  if defined(__OpenBSD__) && OpenBSD >= 201411
+#   if defined(OPENSSL_RAND_SEED_OS)
+#    define OPENSSL_RAND_SEED_GETRANDOM
```

Comment:
It feels pretty weird to be conditionally ```#define```ing ```OPENSSL_RAND_SEED_GETRANDOM``` in the body of a function like this.

Reply:
I didn't like it and had no better idea at that time, but I have now, let me add an other commit.